src/generic.jl

```diff
@@ -1,76 +1,91 @@
-# define abstract types to avoid clash with
-# distances defined for ordinary arrays
-
-# a premetric is a function d that satisfies:
-#
-#   d(x, y) >= 0
-#   d(x, x) = 0
-#
-abstract type ImagePreMetric end
-
-# a semimetric is a function d that satisfies:
-#
-#   d(x, y) >= 0
-#   d(x, x) = 0
-#   d(x, y) = d(y, x)
-#
-abstract type ImageSemiMetric <: ImagePreMetric end
+# TODO: broadcasting
+# TODO: RGB distance
+function colwise!(r::AbstractVector, dist::PreMetric,
```

Comment:
`colwise` is rewritten to incorporate `Distances`

`pairwise` isn't rewritten yet since I find it would make this PR too big(it's too big now)